File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/MopubBannerAdListener.java
package com.monet.bidder;


import com.monet.bidder.threading.InternalRunnable;
import com.mopub.mobileads.DefaultBannerAdListener;
import com.mopub.mobileads.MoPubErrorCode;
import com.mopub.mobileads.MoPubView;

/**
 * Created by nbjacob on 9/2/17.
 */

class MopubBannerAdListener implements MoPubView.BannerAdListener {
    private final static int REFRESH_TRY_DELAY = 4000;
    private final static Logger sLogger = new Logger("BannerAdListener");
    private final String mAdUnitId;
    private final SdkManager mManager;
    private final MoPubView.BannerAdListener mOriginalListener;
    private Runnable mRefreshTimer;

    MopubBannerAdListener(String adUnitId, MoPubView.BannerAdListener originalListener, SdkManager manager) {
        super();
        mAdUnitId = adUnitId;
        mOriginalListener = originalListener == null ? (new DefaultBannerAdListener()) : originalListener;
        mManager = manager;
    }

    void setBannerRefreshTimer(final MoPubView banner) {
        if (mRefreshTimer != null) {
            mManager.getUiThread().run(mRefreshTimer);
        }

        mRefreshTimer = new InternalRunnable() {
            @Override
            public void runInternal() {
                sLogger.debug("Attaching next bid (after load)");
                mManager.addBids(banner, mAdUnitId);
            }

            @Override
            public void catchException(Exception e) {

            }
        };

        mManager.getUiThread().runDelayed(mRefreshTimer, REFRESH_TRY_DELAY);
    }

    @Override
    public void onBannerLoaded(MoPubView banner) {
        sLogger.debug("banner loaded. Attaching next bid");
        setBannerRefreshTimer(banner);
        mOriginalListener.onBannerLoaded(banner);
    }

    @Override
    public void onBannerFailed(MoPubView banner, MoPubErrorCode errorCode) {
        sLogger.debug("banner failed. Attaching new bid");
        mManager.addBids(banner, mAdUnitId);
        mOriginalListener.onBannerFailed(banner, errorCode);
    }

    @Override
    public void onBannerClicked(MoPubView banner) {
        mOriginalListener.onBannerClicked(banner);
    }

    @Override
    public void onBannerExpanded(MoPubView banner) {
        mOriginalListener.onBannerExpanded(banner);
    }

    @Override
    public void onBannerCollapsed(MoPubView banner) {
        mOriginalListener.onBannerCollapsed(banner);
    }
}

File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/MopubBannerAdListener.kt
package com.monet.bidder

import com.monet.bidder.threading.InternalRunnable
import com.mopub.mobileads.DefaultBannerAdListener
import com.mopub.mobileads.MoPubErrorCode
import com.mopub.mobileads.MoPubView
import com.mopub.mobileads.MoPubView.BannerAdListener

/**
 * Created by nbjacob on 9/2/17.
 */
internal class MopubBannerAdListener(
  private val mAdUnitId: String,
  originalListener: BannerAdListener?,
  private val manager: SdkManager
) : BannerAdListener {
  private val mOriginalListener: BannerAdListener = originalListener ?: DefaultBannerAdListener()
  private var mRefreshTimer: Runnable? = null
  fun setBannerRefreshTimer(banner: MoPubView?) {
    if (mRefreshTimer != null) {
      manager.uiThread.run(mRefreshTimer!!)
    }
    mRefreshTimer = object : InternalRunnable() {
      override fun runInternal() {
        sLogger.debug("Attaching next bid (after load)")
        manager.addBids(banner, mAdUnitId)
      }

      override fun catchException(e: Exception?) {}
    }

    manager.uiThread.runDelayed(mRefreshTimer!!, REFRESH_TRY_DELAY.toLong())
  }

  override fun onBannerLoaded(banner: MoPubView) {
    sLogger.debug("banner loaded. Attaching next bid")
    setBannerRefreshTimer(banner)
    mOriginalListener.onBannerLoaded(banner)
  }

  override fun onBannerFailed(
    banner: MoPubView,
    errorCode: MoPubErrorCode
  ) {
    sLogger.debug("banner failed. Attaching new bid")
    manager.addBids(banner, mAdUnitId)
    mOriginalListener.onBannerFailed(banner, errorCode)
  }

  override fun onBannerClicked(banner: MoPubView) {
    mOriginalListener.onBannerClicked(banner)
  }

  override fun onBannerExpanded(banner: MoPubView) {
    mOriginalListener.onBannerExpanded(banner)
  }

  override fun onBannerCollapsed(banner: MoPubView) {
    mOriginalListener.onBannerCollapsed(banner)
  }

  companion object {
    private const val REFRESH_TRY_DELAY = 4000
    private val sLogger = Logger("BannerAdListener")
  }

}